# Vectorization and Numpy

## Stats 141B

## Lecture 6

## Prof. Sharpnack

## Lecture slides at http://anson.ucdavis.edu/~jsharpna/141Blectures/

## Lecture repository at https://github.com/jsharpna/141Blectures/

## Vectorization

- Numpy is matrix algebra package in Python
- Vectorization means converting an algorithm into matrix operations to speed it up
 - Python is slow: it is Just-in-time compiled (JIT)
 - C is fast: it is compiled and variables have static type
 - Vectorization just means that we notice that an operation has a recognizable form and we use the precompiled code for that operation
- basic operation: dot product between $n \times n$ matrices $A,B$

$$ (A B)_{i,j} = \sum_{k=1}^n A_{i,k} B_{k,j} $$

Pseudocode take $O(n^3)$ time:
```
For i = 1,...,n:
  For j = 1,...,n:
    Init C[i,j] = 0
    For k = 1,...,n::
      C[i,j] += A[i,k] B[k,j]
```
* In Python nested for loops are slow, in C they are faster - implement operation in C and import in Python

In [18]:
import numpy as np # our linear alg package

k=10
a = np.arange(k) # built in arrays
type(a), a.shape, a

(numpy.ndarray, (10,), array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]))

In [19]:
a = a+1 # elementwise operation
a

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [20]:
np.exp(a) # elementwise operation

array([2.71828183e+00, 7.38905610e+00, 2.00855369e+01, 5.45981500e+01,
       1.48413159e+02, 4.03428793e+02, 1.09663316e+03, 2.98095799e+03,
       8.10308393e+03, 2.20264658e+04])

In [21]:
## Hypothesis: sum of 1,...,k = k*(k+1) / 2 
## check with array operations
s1 = np.sum(a) # sum the elements
s1 == k * (k+1) / 2

True

In [22]:
## Hypothesis: sum of 1,...,k = k*(k+1) / 2 
## check with only dot products
s1 = a @ np.ones(k) # same as sum
s1 == k * (k+1) / 2

True

$$ \sum_{k=1}^n a_k = \sum_{k=1}^n a_k \cdot 1 = a \bullet \mathbf 1 $$

In [31]:
## Special arrays
onesmat = np.ones((4,4))
zerosmat = np.zeros((4,4))
ident = np.eye(4)

print(onesmat)
print(zerosmat)
print(ident)

## Init with lists of lists
A = np.array([[1,2,3],[5,4,3]])
print(A)
print(A.shape) # shape attribute

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
[[1 2 3]
 [5 4 3]]
(2, 3)


In [43]:
## Slicing arrays

A = np.array([[1,2,3,4],[5,4,3,2],[2,3,4,5]])
print(A[0,2]) # select element
print(A[1:,2]) # select column
print(A[1:,1:]) # select subarray
booind = A[:,0] > 1 # boolean selection
print(A[booind, :])
A[booind,:] = 0. # assignment
print(A)

3
[3 4]
[[4 3 2]
 [3 4 5]]
[[5 4 3 2]
 [2 3 4 5]]
[[1 2 3 4]
 [0 0 0 0]
 [0 0 0 0]]


In [46]:
A_1stk = np.zeros((10,20))
for i,k in enumerate(range(10,20)):
    A_1stk[i,(20-k):] = np.arange(k) + 1

In [47]:
print(A_1stk)

[[ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  1.  2.  3.  4.  5.  6.  7.  8.
   9. 10.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9.
  10. 11.]
 [ 0.  0.  0.  0.  0.  0.  0.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.
  11. 12.]
 [ 0.  0.  0.  0.  0.  0.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11.
  12. 13.]
 [ 0.  0.  0.  0.  0.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12.
  13. 14.]
 [ 0.  0.  0.  0.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13.
  14. 15.]
 [ 0.  0.  0.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14.
  15. 16.]
 [ 0.  0.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15.
  16. 17.]
 [ 0.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16.
  17. 18.]
 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.
  18. 19.]]


In [57]:
s1 = A_1stk.sum(axis=1) # axis to sum along
Ks = np.arange(10,20) # Ks = [10,...,19]
s2 = Ks * (Ks + 1) / 2 # hypothesis
np.array_equal(s1, s2) # testing equality of arrays requires this

True

In [58]:
# The following data is from https://www.pombase.org/downloads/protein-datasets
# It contains the amino acid sequences for proteins in fission yeast

! head ../data/peptide.fa

>SPAC1002.01:pep
MLPPTIRISGLAKTLHIPSRSPLQALKGSFILLNKRKFHYSPFILQEKVQSSNHTIRSDT
KLWKRLLKITGKQAHQFKDKPFSHIFAFLFLHELSAILPLPIFFFIFHSLDWTPTGLPGE
YLQKGSHVAASIFAKLGYNLPLEKVSKTLLDGAAAYAVVKVSYFVENNMVSSTRPFVSN*
>SPAC1002.02:pep
MASTFSQSVFARSLYEDSAENKVDSSKNTEANFPITLPKVLPTDPKASSLHKPQEQQPNI
IPSKEEDKKPVINSMKLPSIPAPGTDNINESHIPRGYWKHPAVDKIAKRLHDQAPSDRTW
SRMVSNLFAFISIQFLNRYLPNTTAVKVVSWILQALLLFNLLESVWQFVRPQPTFDDLQL
TPLQRKLMGLPEGGSTSGKHLTPPRYRPNFSPSRKAENVKSPVRSTTWA*
>SPAC1002.03c:pep


In [59]:
def pep_reader(filename='../data/peptide.fa'):
    with open(filename,'r') as pepfile:
        pepname = False # start of file
        for line in pepfile: 
            if line[0] == '>': # check for prot id line
                if pepname:
                    yield (pepname,pepseq) # if not first output protein
                pepname = line.split(':')[0][1:] # get the id
                pepseq = "" # init seq
            else:
                pepseq += line.strip() # append to seq

In [60]:
pep = pep_reader() # init the iterator
pepdict = {k:v for k,v in enumerate(pep)} # make dictionary with gen expression

In [5]:
def seq_to_set(seq):
    """Protein sequence to a set"""
    return set(p for p in seq)

In [7]:
prot_sets = [seq_to_set(seq) for seq in pepdict.values()] # create sets database

In [64]:
## Creating the alphabet of all amino acids in dataset

alphabet = set()
for seq in prot_sets:
    alphabet |= seq

In [65]:
alphadict = {a:i for i,a in enumerate(alphabet)}
pepnames = [k for k in pepdict.keys()]

In [99]:
print(alphadict) # alphabet - dict
print(pepnames[:3]) # protein names - list
print([pepdict[n] for n in pepnames[:3]]) # sequences - dict

{'H': 0, 'S': 1, 'G': 2, 'N': 3, 'E': 4, 'W': 5, 'T': 6, 'Q': 7, 'I': 8, 'C': 9, 'Y': 10, 'P': 11, 'V': 12, 'R': 13, 'A': 14, 'K': 15, 'D': 16, '*': 17, 'M': 18, 'L': 19, 'F': 20}
['SPAC1002.01', 'SPAC1002.02', 'SPAC1002.03c']
['MLPPTIRISGLAKTLHIPSRSPLQALKGSFILLNKRKFHYSPFILQEKVQSSNHTIRSDTKLWKRLLKITGKQAHQFKDKPFSHIFAFLFLHELSAILPLPIFFFIFHSLDWTPTGLPGEYLQKGSHVAASIFAKLGYNLPLEKVSKTLLDGAAAYAVVKVSYFVENNMVSSTRPFVSN*', 'MASTFSQSVFARSLYEDSAENKVDSSKNTEANFPITLPKVLPTDPKASSLHKPQEQQPNIIPSKEEDKKPVINSMKLPSIPAPGTDNINESHIPRGYWKHPAVDKIAKRLHDQAPSDRTWSRMVSNLFAFISIQFLNRYLPNTTAVKVVSWILQALLLFNLLESVWQFVRPQPTFDDLQLTPLQRKLMGLPEGGSTSGKHLTPPRYRPNFSPSRKAENVKSPVRSTTWA*', 'MRYHGICWFIFQAAIIFAIFGSCQGAFRHQFKTAEQDGFARRNRDLAKFQKENLNWNGLFQLNSISYNSGVVSGVFEQQSENGENQHLFPFSISFLKNDVVRFQMDEKSRLEGTVEYEKNILTKRRFDASTELGFNERAEVYGKDAHLLEQTSTSLTIRYGSHGRFTVIVTFSPFKVEFQRDGEPQVVLNERHLLNMEYYRPKSSRTPEQEANGMWDETFDNFHDSKPKGPESVGLDIKFVDYGNVYGVPEHTSSLSLKETNNSDAGYTEPYRLYNVDLFEYEVDSPMSQYGAIPFMQAHKPNSDVAVFWSNAAATWIDVEKESGPSPHSQSTSTHWYSESGTLDLFIFLGPK

In [100]:
from collections import Counter 

def seq_to_vec(seq,alphadict): 
    """convert amino-acid sequence into array of AA counts"""
    seq_vec = np.zeros(len(alphadict)) # init vector
    seq_count = Counter(seq) # create counter
    for aa, num in seq_count.items(): # amino-acid and count
        seq_vec[alphadict[aa]] = num # store the count in the vector
    return seq_vec

In [101]:
## Build the matrix

AA = np.zeros((len(pepdict), len(alphadict))) # init all zeros
pepind = {} # let's make a reverse lookup dict also
for i,(k,v) in enumerate(pepdict.items()): # i - index, k - name, v - seq
    AA[i,:] = seq_to_vec(v,alphadict) # compute count vector
    pepind[k] = i # update lookup dict

In [102]:
print(AA.shape)
AA

(5136, 21)


array([[ 8., 18.,  8., ...,  2., 24., 15.],
       [ 5., 25.,  6., ...,  4., 21., 10.],
       [31., 70., 62., ..., 14., 67., 59.],
       ...,
       [ 5.,  8.,  5., ...,  2., 16.,  3.],
       [ 4., 11.,  5., ...,  7., 19.,  5.],
       [ 0.,  3.,  0., ...,  1.,  7.,  1.]])

In [103]:
AA[pepind['SPBC16H5.14c'], alphadict['M']]

6.0

### Similarity between proteins

- how similar are the amino acid distribution?
- define the AA proportions (string s, amino acid a):
$$\hat p_s(a) = \frac{\textrm{ # a in s}}{\textrm{ length of s }}$$
- define total variation distance between seqs:
$$ d_T(s_0, s_1) := \sum_{a} |\hat p_{s_0}(a) - \hat p_{s_1}(a)|$$
- define Hellinger distance between seqs:
$$ d_H(s_0, s_1) := \sqrt{ \sum_{a} \left(\sqrt{\hat p_{s_0}(a)} - \sqrt{\hat p_{s_1}(a)}\right)^2 }$$

In [98]:
## compute proportions
seq_lens = AA.sum(axis=1)
AAP = AA / seq_lens.reshape((len(pepind),1)) # broadcasting - replicates the matrix
seq_lens.shape, AA.shape

((5136,), (5136, 21))

### Broadcasting rules

- add 1s to shape (left or right) until match
- replicate right axis until match

``AA / seq_lens.reshape((len(pepind),1))`` - (5136, 21) array / (5136,1) array

In [85]:
AAP

array([[0.04444444, 0.1       , 0.04444444, ..., 0.01111111, 0.13333333,
        0.08333333],
       [0.02173913, 0.10869565, 0.02608696, ..., 0.0173913 , 0.09130435,
        0.04347826],
       [0.03354978, 0.07575758, 0.06709957, ..., 0.01515152, 0.07251082,
        0.06385281],
       ...,
       [0.04201681, 0.06722689, 0.04201681, ..., 0.01680672, 0.13445378,
        0.02521008],
       [0.025     , 0.06875   , 0.03125   , ..., 0.04375   , 0.11875   ,
        0.03125   ],
       [0.        , 0.06976744, 0.        , ..., 0.02325581, 0.1627907 ,
        0.02325581]])

## Computing Hellinger distance
\begin{align*}
d_H(s_0, s_1)^2 &:= \sum_{a} \left(\sqrt{\hat p_{s_0}(a)} - \sqrt{\hat p_{s_1}(a)}\right)^2 \\
 &= \sum_a \hat p_{s_0}(a) + \sum_a \hat p_{s_1}(a) - 2 \sum_a \sqrt{ \hat p_{s_0}(a) \hat p_{s_1}(a) } \\
 &= 2 - 2 \sum_a \sqrt{ \hat p_{s_0}(a) \hat p_{s_1}(a) }
\end{align*}
- can be done with matrix multiply $S_{i_0,j} = \sqrt{p_{s_0}(a)}$ for $i_0$th sequence $s_0$ and jth amino acid $a$
$$ \sum_a \sqrt{ \hat p_{s_0}(a) \hat p_{s_1}(a) } =  (S S^\top)_{i_0,i_1}$$

In [86]:
S = AAP**0.5 # sqrt matrix
Hdist = 2 - 2 * S @ S.T # Hellinger dist matrix

In [87]:
Hdist # Hellinger distance matrix

array([[ 0.00000000e+00,  3.98268445e-02,  6.28791263e-02, ...,
         6.01263096e-02,  1.42239616e-01,  3.08573795e-01],
       [ 3.98268445e-02,  0.00000000e+00,  5.44276215e-02, ...,
         6.59920672e-02,  1.23972206e-01,  3.05000147e-01],
       [ 6.28791263e-02,  5.44276215e-02, -4.44089210e-16, ...,
         7.36722970e-02,  9.04807557e-02,  3.66334820e-01],
       ...,
       [ 6.01263096e-02,  6.59920672e-02,  7.36722970e-02, ...,
         0.00000000e+00,  5.86273341e-02,  2.42683778e-01],
       [ 1.42239616e-01,  1.23972206e-01,  9.04807557e-02, ...,
         5.86273341e-02,  0.00000000e+00,  2.64502164e-01],
       [ 3.08573795e-01,  3.05000147e-01,  3.66334820e-01, ...,
         2.42683778e-01,  2.64502164e-01,  0.00000000e+00]])

In [89]:
Hsind = Hdist.argsort() # sort each row; return indices

In [93]:
## Build the k-Nearest Neighbor graph for Hellinger dist

kNN = {}
k = 10 # k nearest neighbors in Hellinger dist
for pepn, hs in zip(pepnames,Hsind[:,1:k+1]): # zip the protein name with kNNs 
    kNN[pepn] = [pepnames[j] for j in hs] 

In [95]:
kNN['SPBC16H5.14c'] # network as a graph

['SPBC30D10.16',
 'SPBC56F2.08c',
 'SPAC139.01c',
 'SPAC824.02',
 'SPBC336.01',
 'SPAC19G12.12',
 'SPBC776.04',
 'SPBC32H8.03',
 'SPAC1610.02c',
 'SPBC3D6.03c']

## Inference vs. Prediction

- statistical inference: is this effect significant? is the model correct? etc.
- prediction: does this algorithm predict the response variable well?

### terms

- *supervised learning*: predicting one variable from many others
- *predictor* variables: X variables
- *response* variable: Y variable
- ``X``: $n \times p$ design matrix / features
- ``Y``: $n$ label vector


In [104]:
!ls -l ../data/winequality-red.csv

-rw-r--r-- 1 jsharpna jsharpna 84199 Oct  8 10:28 ../data/winequality-red.csv


In [105]:
!head ../data/winequality-red.csv

"fixed acidity";"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"
7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5
7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5
7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;9.8;5
11.2;0.28;0.56;1.9;0.075;17;60;0.998;3.16;0.58;9.8;6
7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5
7.4;0.66;0;1.8;0.075;13;40;0.9978;3.51;0.56;9.4;5
7.9;0.6;0.06;1.6;0.069;15;59;0.9964;3.3;0.46;9.4;5
7.3;0.65;0;1.2;0.065;15;21;0.9946;3.39;0.47;10;7
7.8;0.58;0.02;2;0.073;9;18;0.9968;3.36;0.57;9.5;7


Wine dataset description
- 84199 bytes (not large, feel free to load into memory)
- header with quotations " in the text
- each line has floats without quotations
- each datum separated by ;

In [107]:
with open('../data/winequality-red.csv','r') as winefile:
    header = winefile.readline()
    wine_list = [line.strip().split(';') for line in winefile]

In [108]:
wine_ar = np.array(wine_list,dtype=np.float64)

In [109]:
header

'"fixed acidity";"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"\n'

In [110]:
names = [name.strip('"') for name in header.strip().split(';')]
print(names)

['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality']


In [111]:
#Subselect the predictor X and response y
y = wine_ar[:,-1]
X = wine_ar[:,:-1]
n,p = X.shape

In [112]:
y.shape, X.shape #just checking

((1599,), (1599, 11))

In [113]:
import statsmodels.api as sm

X = np.hstack((np.ones((n,1)),X)) #add intercept
wine_ols = sm.OLS(y,X) #Initialize the OLS 
wine_res = wine_ols.fit()

print(wine_res.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.361
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     81.35
Date:                Mon, 08 Oct 2018   Prob (F-statistic):          1.79e-145
Time:                        10:33:48   Log-Likelihood:                -1569.1
No. Observations:                1599   AIC:                             3162.
Df Residuals:                    1587   BIC:                             3227.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         21.9652     21.195      1.036      0.3

### Linear model

$$\hat y_i = \beta_0 + \sum_{j=1}^p \beta_j x_{i,j}$$

### Inference in linear models

- statistically test for significance of effects
- requires normality assumptions, homoscedasticity, linear model is correct
- hard to obtain significance for individual effect under colinearity

### Prediction perspective

- think of OLS as a black-box model for predicting $Y | X$
- how do we evaluate performance of prediction?
- how do we choose between multiple OLS models?